appsec/go.mod

```diff
@@ -0,0 +1,9 @@
+module github.com/DataDog/dd-trace-go/appsec
```

Comment:
Same comment here. It is not possible for this module to be released and used independently of the main module. To release something new to the user, you'd have to release the main module too with an update to its `go.mod` file pointing to the latest version of this module.

I suggest we keep just the main module, and release it when needed. It's less confusing and more unified to the user too. We should treat this as _the_ Go client. Ultimately we might end merging the metrics repository too.

Reply:
We discussed this approach a while ago with @knusbaum and @felixge, and we agreed that this would be the good option to follow. IIRC, it was not only about letting AppSec release independently, but also to avoid the current issues the tracer go.mod has (the fact it is incomplete to avoid problems with some contrib/ dependencies). There was also the fact the tracer is a gopkg.in module and we could avoid that for a new library by making it a Go module.

I see more benefits than drawbacks so I tend to think this is useful.

Reply:
By the way, there's one big benefit of the no appsec module option that we should be aware of: dyngo can live into dd-trace-go/internal/dyngo so that no one else but us (dd-trace-go/*) can use it. By having separate Go modules, dyngo needs to be public.

Reply:
> All of the packages in this repository are meant to be released, versioned and distributed together. Not separately.

That's where I'm unsure. I was considering that AppSec, Profiling, Debugging, and APM are meant to be usable independently. But on the other hand, we could indeed expose a single client library starting each part according to the configuration.

Maybe that's because I really like modularisation, but even in this latest option, I still prefer the scenario where appsec, apm, profling, debugging are individual go sub-modules, with a "root" module for the client library bundling the others. AFAIK, that's what the ruby team is moving to for example.

Reply:
> That's where I'm unsure. I was considering that AppSec, Profiling, Debugging, and APM would be usable independently. But on the other hand, we could indeed expose a single client library starting each part according to the configuration.

Previously our goal was to have just one client for everything Datadog. I'm not sure if that changed today, but I think not. We should look into it.

> Maybe that's because I really like modularisation, but even in this latest option, I still prefer the scenario where appsec, apm, profling, debugging are individual go sub-modules, with a "root" module for the client library bundling the others. 

I agree with this conceptually, it would be nice. If we go this way, we should probably do it globally as a concept which applies to the entire repository (as you have mentioned yourself), for all products included. Doing it just for one and leaving the rest bundled is just inconsistent.

But the main idea here remains that from a user-perspective there is no AppSec module, package, or API. It's just new functionality applied on top of already existing APM integrations. Am I wrong in understanding here?

Reply:
> But the main idea here remains that from a user-perspective there is no AppSec module, package, or API. It's just new functionality applied on top of already existing APM integrations. Am I wrong in understanding here?

As of now yes. The longer term idea was to rely on dynamic instrumentation applying (dyngo) integrations APM/AppSec could listen to.

Reply:
Then I think that we are all on the same page here that we do not need sub-modules or public packages. Placing all of it inside the root `internal` package will do. We can then later on evaluate next steps and approaches when new requirements arise and we want to release public API.

Reply:
Based on the thoughtful discussion above, I support @gbbr's suggestion to not create a new go module for appsec in this PR.

I am very interested in exploring this pattern, but since appsec is coupled to the tracer for now, and both need to be released at the same time, I don't see the immediate benefit.

Perhaps more importantly, I'm worried that having a lengthy discussion about the multi module pattern would distract from reviewing the actual contents of this PR. This PR is too big and complex to risk that IMO.

Additionally I agree that as much of the new API as possible should be `internal`.

Reply:
Sounds good to me too.
The module is now removed (2259fa6dd2c9ad6c2a310125f68a2e91a0e011df) and results into this split:
- `internal/appsec`
- `internal/dyngo`

Reply:
Reopening the conversation just to let @knusbaum know about it :-)